File: src/Syntax/JavaFileFactory.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Net;
using System.Text;
using AutoRest.Core;
using AutoRest.Core.Utilities;
using AutoRest.Extensions;
using AutoRest.Core.Model;
using Newtonsoft.Json;
using AutoRest.Core.Utilities.Collections;
using Newtonsoft.Json.Linq;
using AutoRest.Java.Model;
using System.IO;

namespace AutoRest.Java
{
    public class JavaFileFactory
    {
        private JavaSettings settings;

        public JavaFileFactory(JavaSettings settings)
        {
            this.settings = settings;
        }

        public JavaFile CreateEmptySourceFile(string package, string fileNameWithoutExtension)
        {
            string folderPath = Path.Combine("src", "main", "java", package.Replace('.', Path.DirectorySeparatorChar));
            string filePath = Path.Combine(folderPath, $"{fileNameWithoutExtension}.java")
                .Replace('\\', '/')
                .Replace("//", "/");
            return new JavaFile(filePath);
        }

        public JavaFile CreateSourceFile(string package, string fileNameWithoutExtension)
        {
            JavaFile javaFile = CreateEmptySourceFile(package, fileNameWithoutExtension);

            string headerComment = settings.FileHeaderText;
            if (!string.IsNullOrEmpty(headerComment))
            {
                javaFile.LineComment(settings.MaximumJavadocCommentWidth, (comment) =>
                {
                    comment.Line(headerComment);
                });
                javaFile.Line();
            }

            javaFile.Package(package);
            javaFile.Line();

            return javaFile;
        }

        public JavaFile CreateTestFile(string package, string fileNameWithoutExtension)
        {
            string folderPath = Path.Combine("src", "test", "java", package.Replace('.', Path.DirectorySeparatorChar));
            string filePath = Path.Combine(folderPath, $"{fileNameWithoutExtension}.java")
                .Replace('\\', '/')
                .Replace("//", "/");
            JavaFile javaFile = new JavaFile(filePath);

            string headerComment = settings.FileHeaderText;
            if (!string.IsNullOrEmpty(headerComment))
            {
                javaFile.JavadocComment(settings.MaximumJavadocCommentWidth, (comment) =>
                {
                    comment.Description(headerComment);
                });
                javaFile.Line();
            }

            javaFile.Package(package);
            javaFile.Line();

            return javaFile;
        }
    }
}

File: mapper/src/main/java/com/azure/autorest/model/javamodel/JavaFileFactory.java
package com.azure.autorest.model.javamodel;

import AutoRest.Core.*;
import AutoRest.Core.Utilities.*;
import AutoRest.Extensions.*;
import AutoRest.Core.Model.*;
import Newtonsoft.Json.*;
import AutoRest.Core.Utilities.Collections.*;
import Newtonsoft.Json.Linq.*;
import AutoRest.Java.Model.*;
import java.io.*;
import java.nio.file.*;

// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.


public class JavaFileFactory
{
    private JavaSettings settings;

    public JavaFileFactory(JavaSettings settings)
    {
        this.settings = settings;
    }

    public final JavaFile CreateEmptySourceFile(String package_Keyword, String fileNameWithoutExtension)
    {
        String folderPath = Path.Combine("src", "main", "java", package_Keyword.replace('.', File.separatorChar));
        String filePath = Paths.get(folderPath).resolve(String.format("%1$s.java", fileNameWithoutExtension)).toString().Replace('\\', '/').replace("//", "/");
        return new JavaFile(filePath);
    }

    public final JavaFile CreateSourceFile(String package_Keyword, String fileNameWithoutExtension)
    {
        JavaFile javaFile = CreateEmptySourceFile(package_Keyword, fileNameWithoutExtension);

        String headerComment = settings.getFileHeaderText();
        if (headerComment != null && !headerComment.isEmpty())
        {
            javaFile.LineComment(settings.getMaximumJavadocCommentWidth(), (comment) ->
            {
                    comment.Line(headerComment);
            });
            javaFile.Line();
        }

        javaFile.Package(package_Keyword);
        javaFile.Line();

        return javaFile;
    }

    public final JavaFile CreateTestFile(String package_Keyword, String fileNameWithoutExtension)
    {
        String folderPath = Path.Combine("src", "test", "java", package_Keyword.replace('.', File.separatorChar));
        String filePath = Paths.get(folderPath).resolve(String.format("%1$s.java", fileNameWithoutExtension)).toString().Replace('\\', '/').replace("//", "/");
        JavaFile javaFile = new JavaFile(filePath);

        String headerComment = settings.getFileHeaderText();
        if (headerComment != null && !headerComment.isEmpty())
        {
            javaFile.JavadocComment(settings.getMaximumJavadocCommentWidth(), (comment) ->
            {
                    comment.Description(headerComment);
            });
            javaFile.Line();
        }

        javaFile.Package(package_Keyword);
        javaFile.Line();

        return javaFile;
    }
}